### PR TITLE
Implementation of Shannon Entropy for Random Ray

### LINKED ISSUE #2868: Shannon Entropy for Random Ray
Similar to Monte Carlo, computation of Shannon Entropy is a useful way to determine if enough inactive batches have been run such that the source has fully developed and become a stationary distribution.

In Monte Carlo, the Shannon Entropy is calculated using the number of fission site particles that are sampled from each cell within an overlaid Cartesian mesh. However, with random ray, as we already have a well refined spatial mesh (due to the need for FSRs), it is likely sufficient to compute the Shannon Entropy using the source terms of each FSR directly. I.e., there is no need to define a secondary spatial mesh. Additionally, as the Shannon entropy can be trivially deduced from the source, it may make sense to always compute it rather than requiring a user input setting to be turned on.

Note that Shannon Entropy will also likely be useful when running in fixed source mode, as in random ray inactive batches are still needed to converge the scattering source.

### PR DESCRIPTION
**Pull request by ethankrammer**
# Description

Shannon entropy is a method of diagnosing convergence for Monte Carlo eigenvalue problems. Prior to this PR, the Shannon entropy was only being computed in multigroup MC and not Random Ray. Since Random Ray also needs to converge the scattering source in addition to the eigenvalue, there is additional need for computing Shannon entropy in the Random Ray method.

Since the Random Ray implementation in OpenMC uses Flat Source Regions, we use these instead of a user-defined mesh to compute Shannon entropy. The implementation otherwise follows the same methodology for [calculating Shannon entropy as other MC applications](https://www.oecd-nea.org/science/wpncs/sccsa/documents/brown-physor-2006.pdf).

A regression test directory, `random_ray_entropy`, was also developed for this PR. Based on the cross-sections and geometry mesh [given by Nowak et al.](https://doi.org/10.1016/j.anucene.2016.05.002), the Shannon entropy is calculated exactly as 9.0 and the eigenvalue as 1.0. 

Small edits to the documentation were also made to reflect Shannon entropy being implemented into the Random Ray method in OpenMC.

The [2D C5G7 Random ray example ](https://github.com/jtramm/trrm_examples/tree/main/2D_c5g7) was used for validation of the Shannon entropy implementation. Below is a plot of the eigenvalue and Shannon entropy calculated at each batch.

2D C5G7 Shannon Entropy Convergence

Fixes #2868

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by ethankrammer** [comment]
This PR has a known issue where Shannon entropy is not calculated correctly for problems with flat source regions (FSRs) of uneven volume. The test regression test, `random_ray_entropy`, runs as expected, since the lattice builds FSRs of equal volume.

While calculating the "correct" value of Shannon entropy is not particularly useful, this issue may result in inaccurate convergence diagnostics for problems with large FSR volume differences. Essentially, the smaller FSRs dominate the calculation of Shannon entropy, which may result in perceived convergence even if the larger FSRs have yet to converge.

File: src/random_ray/flat_source_domain.cpp [hunk 1]
```diff
@@ -247,12 +252,39 @@ double FlatSourceDomain::compute_k_eff(double k_eff_old) const
...
+    // Temporary FSR fission rate for performance.
+    double temp_sr_fission_rate = sr_fission_source_new * volume;
+
+    // Calculating old and new fission rates.
     fission_rate_old += sr_fission_source_old * volume;
-    fission_rate_new += sr_fission_source_new * volume;
+    fission_rate_new += temp_sr_fission_rate;
+
+    // Assigning the fission source to the vector for entropy calculation.
+    p[sr] = temp_sr_fission_rate;
```
Comment (jtramm) [inline src/random_ray/flat_source_domain.cpp:263]: Might be simpler to cut the temporary variable and just do something like
```C++
// Compute total fission rates in FSR
sr_fission_source_old *= volume;
sr_fission_source_new *= volume;

// Accumulate totals
fission_rate_old += sr_fission_source_old;
fission_rate_new += sr_fission_source_new;

// Store total fission rate in the FSR for Shannon calculation
p[sr] = sr_fission_source_new;

```

File: src/random_ray/flat_source_domain.cpp [hunk 2]
```diff
@@ -247,12 +252,39 @@ double FlatSourceDomain::compute_k_eff(double k_eff_old) const
...
+  const double inv_log2 = 1.0 / std::log(2.0);
+
+#pragma omp parallel for reduction(+ : H)
+  for (int i = 0; i < n_source_regions_; i++) {
+    // Only if FSR has fission source
+    if (p[i] != 0.0f) {
+      // Normalize to total weight of bank sites. p_i for better performance
+      float p_i = p[i] * inverse_sum;
+      // Sum values to obtain Shannon entropy.
+      H -= p_i * std::log(p_i) * inv_log2;
```
Comment (jtramm) [inline src/random_ray/flat_source_domain.cpp:281]: Is there a reason not to do the below?
```suggestion
      H -= p_i * std::log2(p_i);
```

File: src/random_ray/flat_source_domain.cpp [hunk 3]
```diff
@@ -247,12 +252,39 @@ double FlatSourceDomain::compute_k_eff(double k_eff_old) const
...
+  double H = 0.0;
+  // defining an inverse sum for better performance
+  double inverse_sum = 1 / fission_rate_new;
+  // defining inverse log(2) for better performance
+  const double inv_log2 = 1.0 / std::log(2.0);
+
+#pragma omp parallel for reduction(+ : H)
+  for (int i = 0; i < n_source_regions_; i++) {
+    // Only if FSR has fission source
+    if (p[i] != 0.0f) {
```
Comment (jtramm) [inline src/random_ray/flat_source_domain.cpp:277]: Slightly negative fluxes can happen occasionally (due to the simulation averaged volume estimator), which might cause H to go `NaN`. Might be better to check for greater than zero?
```suggestion
    if (p[i] > 0.0f) {
```

File: src/settings.cpp [hunk 4]
```diff
@@ -624,29 +624,36 @@ void read_settings_xml(pugi::xml_node root)
   }
 
   // Shannon Entropy mesh
-  if (check_for_node(root, "entropy_mesh")) {
-    int temp = std::stoi(get_node_value(root, "entropy_mesh"));
-    if (model::mesh_map.find(temp) == model::mesh_map.end()) {
-      fatal_error(fmt::format(
-        "Mesh {} specified for Shannon entropy does not exist.", temp));
+  if (solver_type == SolverType::RANDOM_RAY) {
```
Comment (jtramm) [inline src/settings.cpp:627]: Might change the `// Shannon Entropy mesh` comment to just `// Shannon Entropy` as this section is doing more now.

**Review by jtramm — CHANGES_REQUESTED** [review]
Nice work @ethankrammer! This appears to be working nicely. I've made a few minor comments and requests for small changes, and then this should be in good shape to go in.

File: src/random_ray/flat_source_domain.cpp [hunk 5]
```diff
@@ -225,6 +226,10 @@ double FlatSourceDomain::compute_k_eff(double k_eff_old) const
   const int t = 0;
   const int a = 0;
 
+  // Vector for gathering fission source terms for Shannon entropy calculation
+  vector<float> p(n_source_regions_, 0.0f);
+  double sum = 0.0;
```
[the remark this answers is not in the document]
Comment (ethankrammer) [reply src/random_ray/flat_source_domain.cpp:231]: It is not. Thanks!

File: src/random_ray/flat_source_domain.cpp (hunk 1 above)
Comment (ethankrammer) [reply src/random_ray/flat_source_domain.cpp:263]: I like that approach, thanks for the suggestion!

File: src/random_ray/flat_source_domain.cpp (hunk 2 above)
Comment (ethankrammer) [reply src/random_ray/flat_source_domain.cpp:281]: Regarding our discussion, I've changed this and the normal MC Shannon entropy to log2.

File: src/random_ray/flat_source_domain.cpp (hunk 3 above)
Comment (ethankrammer) [reply src/random_ray/flat_source_domain.cpp:277]: Great to know, I've gone ahead and changed this.

File: src/settings.cpp (hunk 4 above)
Comment (ethankrammer) [reply src/settings.cpp:627]: Changed, thanks!

File: docs/source/usersguide/random_ray.rst [hunk 6]
```diff
@@ -46,7 +46,10 @@ magnitude or more. For instance, it may be reasonable to only use 50 inactive
 batches for a light water reactor simulation with Monte Carlo, but random ray
 might require 500 or more inactive batches. Similar to Monte Carlo,
 :ref:`Shannon entropy <usersguide_entropy>` can be used to gauge whether the
-combined scattering and fission source has fully developed.
+combined scattering and fission source has fully developed. The Shannon entropy
+is calculated automatically near the end of each batch and is printed to the
```
[the remark this answers is not in the document]
Comment (ethankrammer) [reply docs/source/usersguide/random_ray.rst:50]: That does make more sense. When I was writing this, I was thinking of a "batch," as the full process (including running the batch and running functions like `finalize_generation()` and `finalize_batch()`. I've changed this to reflect your input.

File: docs/source/usersguide/random_ray.rst [hunk 7]
```diff
@@ -46,7 +46,10 @@ magnitude or more. For instance, it may be reasonable to only use 50 inactive
 batches for a light water reactor simulation with Monte Carlo, but random ray
 might require 500 or more inactive batches. Similar to Monte Carlo,
 :ref:`Shannon entropy <usersguide_entropy>` can be used to gauge whether the
-combined scattering and fission source has fully developed.
+combined scattering and fission source has fully developed. The Shannon entropy
```
[the remark this answers is not in the document]
Comment (ethankrammer) [reply docs/source/usersguide/random_ray.rst:49]: This was an oversight on my part, thank you for the suggestion!

File: docs/source/methods/random_ray.rst [hunk 8]
```diff
@@ -733,6 +733,30 @@ How are Tallies Handled?
...
+The Shannon entropy is then computed normally as
+
+.. math::
+    :label: shannon-entropy-random-ray
+
+    H = - \sum_{i=1}^N S_i \log_2 S_i
+
+where :math:`N` is the number of FSRs. FSRs with no fission source sites are skipped
+to avoid taking the logarithm of zero in :eq:`shannon-entropy-random-ray`.
+
```
[the remark this answers is not in the document]
Comment (ethankrammer) [reply docs/source/methods/random_ray.rst:759]: Thank you for this input! I have made these edits.

[the remark this answers is not in the document]
Comment (jtramm) [reply docs/source/methods/random_ray.rst]: This is getting closer! A few notes on the revised docs:

- The updated equation still doesn't seem to match the code. The numerator should be summing over all groups in the FSR, and the denominator should be summing up the total fission source in the simulation (e.g., all groups and FSRS).  Computation of the total fission source in the denominator will also involve the volume. Also -- the $\times$ symbol is fine for use with text based equations, but when making a purely symbolic math equation it's uncommon to see a $\times$ symbol.

- The phrasing in the scattering source section could use some more work: "Shannon entropy is calculated differently for the random ray solver, as described in ..., as the Shannon entropy does not serve as a diagnostic tool for convergence of the scattering source." For k-eff calculations, both solvers are using the fission source for the convergence criteria, and the difference is in what sort of mesh is being used (user-supplied vs. random ray flat source regions). Neither solver is using the scattering source as input into the Shannon entropy calculation, but the way the sentence is written implies that Shannon Entropy is being used in Monte Carlo to diagnose convergence of the scattering source. Why mention the scattering source at all?

File: docs/source/usersguide/random_ray.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (jtramm) [reply docs/source/usersguide/random_ray.rst]: "However, unlike Monte Carlo, the Shannon entropy computed for random ray does not show convergence of the scattering source." As far as I know, Shannon entropy is not used to show convergence of the scattering source in MC? The scattering source is implicitly converged every iteration as the particle history is tracked through all its scatters from birth -> death/escape, so one does not need to measure scattering source convergence in MC.

File: src/random_ray/flat_source_domain.cpp [hunk 9]
```diff
@@ -247,12 +251,38 @@ double FlatSourceDomain::compute_k_eff(double k_eff_old) const
...
   }
 
   double k_eff_new = k_eff_old * (fission_rate_new / fission_rate_old);
 
+  double H = 0.0;
+  // defining an inverse sum for better performance
+  double inverse_sum = 1 / fission_rate_new;
+
+#pragma omp parallel for reduction(+ : H)
+  for (int i = 0; i < n_source_regions_; i++) {
```
Comment (jtramm) [inline src/random_ray/flat_source_domain.cpp:273]: I'm noticing that we have two for loops in this function doing the same loop over all source regions, but are using two different iterator variable names (`sr` vs. `i`). I think we tend to use `sr` pretty consistently for looping over source regions through the code, so might as well use that for the second loop as well.

Comment (ethankrammer) [reply src/random_ray/flat_source_domain.cpp]: This is a good convention! I have made this edit.

**Review by jtramm — APPROVED** [review]

**Comment by jtramm** [comment]
The changes all look good -- thanks @ethankrammer, and congrats on your first OpenMC PR!

**Pull request merged**